docs/component-stability.md

```diff
@@ -66,6 +66,74 @@ Stable components MUST be compatible between minor versions unless critical secu
 component owner MUST provide a migration path and a reasonable time frame for users to upgrade. The same rules from beta
 components apply to stable when it comes to configuration changes.
 
+#### Observability requirements
+
+Stable components should emit enough internal telemetry to let users detect errors, as well as data
+loss and performance issues inside the component, and to help diagnose them if possible.
+
+The internal telemetry of a stable component should allow observing the following:
+
+1. How much data the component receives.
+
+  For receivers, this could be a metric counting requests, received bytes, scraping attempts, etc.
+
+  For other components, this would typically be the number of items (log records, metric points,
+  spans) received through the `Consumer` API.
+
+2. How much data the component outputs.
+
+  For exporters, this could be a metric counting requests, sent bytes, etc.
+
+  For other components, this would typically be the number of items forwarded through the `Consumer`
+  API.
+
+3. How much data is dropped because of errors.
+
+  For receivers, this could include a metric counting payloads that could not be parsed in.
+  
+  For receivers and exporters, this could include a metric counting requests that failed because
+  of network errors.
+
+  The goal is to be able to easily pinpoint the source of data loss in the Collector pipeline, so
+  this should either:
+  - only include errors internal to the component, or;
+  - allow distinguishing said errors from ones originating in an external service, or propagated
+    from downstream Collector components.
+
+4. Details for error conditions.
+
+  This could be in the form of logs or spans detailing the reason for an error. As much detail as
+  necessary should be provided to ease debugging. Processed signal data should not be included for
+  security and privacy reasons.
+
+5. Other discrepancies between input and output. This may include:
+
+  - How much data is dropped as part of normal operation (eg. filtered out).
+
+  - How much data is created by the component.
+
+  - How much data is currently held by the component (eg. an UpDownCounter keeping track of the
+    size of an internal queue).
+
+6. Processing performance.
+
+  This could be a histogram of end-to-end component latency, measured as the time between external
+  requests or `Consumer` API calls.
```

Comment:
I feel like this is something that would be capture by the RED metrics from a component like the [spanmetrics connector](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/connector/spanmetricsconnector#overview). Should we instead require that the main loop of the component is be traced?

Reply:
If by "main loop" you mean everything from after the component entry point (HTTP/gRPC/whatever server for receivers, `Consume` for others) to before the exit point (HTTP/gRPC/whatever request for exporters, `Consume` for others), I think it would probably be doable. The requirement cites histograms as an example, so this would not technically be excluded.

I avoided requiring spans because the PI RFC was pretty frilly about them and preferred logs when metrics are insufficient. If we're considering using spans again, we may also want to reconsider outputting errors as logs and encourage emitting them as span events instead.